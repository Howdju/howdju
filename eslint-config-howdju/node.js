/** ESLint config for node code: either build or backend code. */

const { testFilePattern } = require("./constants");

module.exports = {
  env: {
    node: true,
  },
  plugins: ["promise"],
  extends: ["eslint-config-howdju", "plugin:promise/recommended", "prettier"],
  rules: {
    // This incorrectly warns for Bluebird.catch with 2+ arguments (which is correct invocation.)
    "promise/valid-params": "off",
  },
  overrides: [
    {
      files: testFilePattern,
      rules: {
        // Tests don't need to catch errors.  If we are asserting properly, then the tests should
        // fail if the error causes us to miss the assertion
        "promise/catch-or-return": "off",
      },
    },
  ],
};
