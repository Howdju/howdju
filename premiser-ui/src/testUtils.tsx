import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { createMemoryHistory, History, createLocation } from "history";
import { persistStore, PersistorOptions } from "redux-persist";
import { compile } from "path-to-regexp";
import { match } from "react-router";
import userEvent from "@testing-library/user-event";
import { setupServer } from "msw/node";
import { Saga } from "redux-saga";
import { head } from "lodash";

import { AppStore, RootState, sagaMiddleware, setupStore } from "./setupStore";

interface ProviderRenderOptions
  extends DefaultStoreOptions,
    Omit<RenderOptions, "queries"> {
  persist?: boolean;
}

export interface DefaultStoreOptions {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
  history?: History<any>;
}

export function setupDefaultStore({
  preloadedState = {},
  history = createMemoryHistory(),
  store = setupStore(history, preloadedState),
}: DefaultStoreOptions = {}) {
  return { preloadedState, history, store };
}

/**
 * Use Jest fake timers.
 *
 * Restores real timers after each to allow third-party libraries to cleanup. See
 * https://testing-library.com/docs/using-fake-timers/
 *
 * TODO(219): should we do this in a Jest environment instead?
 */
export function withFakeTimers() {
  beforeEach(() => {
    // Use fake timers so that we can ensure animations complete before snapshotting.
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });
}

/** Configures an msw fake server for the test.
 *
 * @returns the mock server
 */
export function withMockServer() {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });
  afterEach(() => {
    server.resetHandlers();
  });
  afterAll(() => {
    server.close();
  });
  return server;
}

/** Render a React component with the redux store etc. */
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    history = createMemoryHistory(),
    store = setupStore(history, preloadedState),
    persist = true,
    ...renderOptions
  }: ProviderRenderOptions = {}
) {
  const persistor = persistStore(store, {
    manualPersist: true,
  } as PersistorOptions);
  if (persist) {
    persistor.persist();
  }
  function Wrapper({
    children,
  }: PropsWithChildren<Record<string, unknown>>): JSX.Element {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  }

  return {
    store,
    persistor,
    history,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export interface MakeRouteComponentPropsOptions<T> {
  pathParams?: T;
  searchParams?: Record<string, string>;
}

/**
 * Makes RouteComponentProps.
 *
 * @param path The parameterized react-router path: `/p/:id`.
 * @param params the parameters to substitute in the path: `{id: "123"}`.
 */
export function makeRouteComponentProps<T extends Record<string, string>>(
  path: string,
  options?: MakeRouteComponentPropsOptions<T>
) {
  const { pathParams = {} as T, searchParams = {} } = options ?? {};
  const toUrl = compile(path, { encode: encodeURIComponent });
  const url = toUrl(pathParams);
  const match: match<T> = {
    isExact: false,
    path,
    url,
    params: pathParams,
  };
  const locationPath = searchParams
    ? match.url + "?" + new URLSearchParams(searchParams)
    : match.url;
  const location = createLocation(locationPath);

  return { match, location };
}

/**
 * Filters the elements to only those that are visible.
 *
 * react-md uses a hidden and visible input element as an implementation detail. In our tests we
 * often only want to assert on the visible one.
 */
export function ariaVisibleOne(
  elements: HTMLElement[]
): HTMLElement | undefined {
  return head(
    elements.filter((e) => e.getAttribute("aria-hidden") !== "false")
  );
}

/** Run a saga with the redux store etc. in place. */
export async function testSaga(saga: Saga<any[]>) {
  const history = createMemoryHistory();
  const preloadedState = {};
  const store = setupStore(history, preloadedState);
  const persistor = persistStore(store, {
    manualPersist: true,
  } as PersistorOptions);
  persistor.persist();
  await sagaMiddleware.run(saga).toPromise();
}

export function setupUserEvent() {
  // userEvent delays between actions, so make sure it advances the fake timers.
  // (https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841)
  return userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
}
