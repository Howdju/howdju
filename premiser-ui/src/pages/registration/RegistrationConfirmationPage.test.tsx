import React from "react";
import { rest } from "msw";
import { screen } from "@testing-library/react";
import { setupServer } from "msw/node";
import userEvent from "@testing-library/user-event";

import RegistrationConfirmationPage from "./RegistrationConfirmationPage";
import { renderWithProviders } from "@/testUtils";
import {
  CreateRegistrationConfirmation,
  httpStatusCodes,
  ErrorOut,
} from "howdju-common";

const server = setupServer();

beforeAll(() => {
  server.listen();
});
beforeEach(() => {
  // Use fake timers so that we can ensure animations complete before snapshotting.
  jest.useFakeTimers();
});
afterEach(() => {
  server.resetHandlers();
  jest.runOnlyPendingTimers();
  jest.useRealTimers();
});
afterAll(() => server.close());

describe("RegistrationConfirmationPage", () => {
  test("shows form", async () => {
    // Arrange
    const registrationCode = "fake-registration-code";
    const email = "the-email@domain.com";
    server.use(
      rest.get("http://localhost/registration-requests", (_req, res, ctx) => {
        return res(ctx.status(httpStatusCodes.OK), ctx.json({ email }));
      })
    );

    // Act
    const { container, history } = renderWithProviders(
      <RegistrationConfirmationPage />
    );
    history.push(`complete-registration?registrationCode=${registrationCode}`);
    await screen.findByLabelText(/username/i);

    // Assert
    expect(screen.getByRole("heading")).toHaveTextContent(
      /complete registration/i
    );
    expect(
      screen.getByRole("button", { name: /complete registration/i })
    ).toHaveClass("md-btn--raised-disabled");
    expect(container).toMatchSnapshot();
  });

  test("shows error message for missing registration code", async () => {
    server.use(
      rest.get("http://localhost/registration-requests", (_req, res, ctx) => {
        return res(ctx.status(httpStatusCodes.NOT_FOUND));
      })
    );

    const { container } = renderWithProviders(<RegistrationConfirmationPage />);
    await screen.findByRole("heading");

    expect(screen.getByText(/missing registration code/i)).toBeInTheDocument();
    jest.runAllTimers();
    expect(container).toMatchSnapshot();
  });

  test("shows error message for duplicate username", async () => {
    // Arrange

    // userEvent delays between actions, so make sure it advances the fake timers.
    // (https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841)
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

    const username = "the_username";

    server.use(
      rest.get("http://localhost/registration-requests", (_req, res, ctx) => {
        return res(
          ctx.status(httpStatusCodes.OK),
          ctx.json({ email: "a@b.com" })
        );
      }),

      rest.post("http://localhost/registrations", (_req, res, ctx) => {
        const response: ErrorOut<{
          registrationConfirmation: CreateRegistrationConfirmation;
        }> = {
          errorCode: "VALIDATION_ERROR",
          errors: {
            _errors: [],
            registrationConfirmation: {
              _errors: [],
              username: {
                _errors: [
                  {
                    code: "custom",
                    path: ["username"],
                    message: `The username is already in use: ${username}`,
                    params: { code: "ALREADY_IN_USE", username },
                  },
                ],
              },
            },
          },
        };
        return res(ctx.status(httpStatusCodes.CONFLICT), ctx.json(response));
      })
    );

    const { container, history } = renderWithProviders(
      <RegistrationConfirmationPage />
    );
    history.push(`complete-registration?registrationCode=abc`);
    await screen.findByLabelText(/username/i);

    await user.type(screen.getByLabelText(/username/i), username);
    await user.type(screen.getByLabelText(/password/i), "password");
    await user.type(screen.getByLabelText(/full name/i), "Testius Maximus IV");
    await user.type(screen.getByLabelText(/first name/i), "Testy");
    await user.click(
      screen.getByRole("checkbox", {
        name: /read and agree to the User Agreement/i,
      })
    );
    await user.click(
      screen.getByRole("checkbox", { name: /13 years old or older/i })
    );
    await user.click(
      screen.getByRole("checkbox", {
        name: /consent to the processing of my personal data/i,
      })
    );
    await user.click(
      screen.getByRole("checkbox", {
        name: /I am not located in the European Union/i,
      })
    );
    expect(
      screen.getByRole("button", { name: /complete registration/i })
    ).toBeEnabled();

    // Act
    await user.click(
      screen.getByRole("button", { name: /complete registration/i })
    );

    // Assert
    expect(
      await screen.findByText(
        new RegExp(`the username is already in use: ${username}`, "i")
      )
    ).toBeInTheDocument();
    // Let all animations complete before taking the snapshot. Otherwise we get flaky UI
    // differences depending on where the material UI animations are at.
    jest.runAllTimers();
    expect(container).toMatchSnapshot();
  });

  test("shows success message upon confirmation.", async () => {
    // Arrange

    // userEvent delays between actions, so make sure it advances the fake timers.
    // (https://github.com/testing-library/user-event/issues/833#issuecomment-1171452841)
    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

    const registrationCode = "abc";
    const username = "the_username";

    server.use(
      rest.get("http://localhost/registration-requests", (req, res, ctx) => {
        const requestRegistrationCode =
          req.url.searchParams.get("registrationCode");
        if (requestRegistrationCode !== registrationCode) {
          return res(ctx.status(httpStatusCodes.NOT_FOUND));
        }
        return res(
          ctx.status(httpStatusCodes.OK),
          ctx.json({ email: "a@b.com" })
        );
      }),
      rest.post("http://localhost/registrations", (_req, res, ctx) => {
        return res(ctx.status(httpStatusCodes.OK));
      })
    );

    const { container, history } = renderWithProviders(
      <RegistrationConfirmationPage />
    );
    history.push(`complete-registration?registrationCode=${registrationCode}`);
    await screen.findByLabelText(/username/i);

    await user.type(screen.getByLabelText(/username/i), username);
    await user.type(screen.getByLabelText(/password/i), "password");
    await user.type(screen.getByLabelText(/full name/i), "Testius Maximus IV");
    await user.type(screen.getByLabelText(/first name/i), "Testy");
    await user.click(
      screen.getByRole("checkbox", {
        name: /read and agree to the User Agreement/i,
      })
    );
    await user.click(
      screen.getByRole("checkbox", { name: /13 years old or older/i })
    );
    await user.click(
      screen.getByRole("checkbox", {
        name: /consent to the processing of my personal data/i,
      })
    );
    await user.click(
      screen.getByRole("checkbox", {
        name: /I am not located in the European Union/i,
      })
    );
    expect(
      screen.getByRole("button", { name: /complete registration/i })
    ).toBeEnabled();

    // Act
    await user.click(
      screen.getByRole("button", { name: /complete registration/i })
    );

    // Assert
    expect(
      await screen.findByText(/You are now logged in/i)
    ).toBeInTheDocument();
    expect(
      screen.queryByText(
        /Please enter the following to complete your registration/i
      )
    ).not.toBeInTheDocument();
    // Let all animations complete before taking the snapshot. Otherwise we get flaky UI
    // differences depending on where the material UI animations are at.
    jest.runAllTimers();
    expect(container).toMatchSnapshot();
  });
});
