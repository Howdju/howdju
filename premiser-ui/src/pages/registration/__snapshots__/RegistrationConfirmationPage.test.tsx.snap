// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationConfirmationPage shows error message for duplicate username 1`] = `
<div>
  <div
    id="register-page"
  >
    <div
      class="rmd-grid"
      style="--rmd-grid-cols: 1;"
    >
      <div
        class="rmd-card rmd-card--shadowed rmd-grid__cell"
      >
        <header
          class="rmd-card__header rmd-card__header--center"
        >
          <span
            class="rmd-card__header-content"
          >
            <h5
              class="rmd-card__title"
            >
              Complete Registration
            </h5>
          </span>
        </header>
        <div
          class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
        >
          <div
            class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
          >
            Please enter the following to complete your registration
          </div>
          <form>
            <div
              class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
            >
              <ul
                class="error-message"
              />
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--disabled rmd-text-field-container--label rmd-text-field-container--outline-left"
                  style="flex-grow: 1;"
                >
                  <span
                    class="rmd-text-field-addon rmd-text-field-addon--presentational"
                  >
                    <span
                      class="material-symbols"
                    >
                      alternate_email
                    </span>
                  </span>
                  <label
                    class="rmd-label rmd-label--disabled rmd-floating-label rmd-floating-label--active"
                    for="email"
                  >
                    Email
                  </label>
                  <input
                    autocomplete="email"
                    class="rmd-text-field rmd-text-field--floating"
                    disabled=""
                    id="email"
                    type="email"
                    value="a@b.com"
                  />
                </div>
              </div>
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--error rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--outline-error rmd-text-field-container--outline-left"
                  style="flex-grow: 1;"
                >
                  <span
                    class="rmd-text-field-addon rmd-text-field-addon--presentational"
                  >
                    <span
                      class="material-symbols"
                    >
                      person
                    </span>
                  </span>
                  <label
                    class="rmd-label rmd-label--error rmd-floating-label rmd-floating-label--active"
                    for="username"
                  >
                    Username
                  </label>
                  <input
                    autocomplete="username"
                    class="rmd-text-field rmd-text-field--floating"
                    id="username"
                    maxlength="64"
                    minlength="3"
                    name="username"
                    required=""
                    type="text"
                    value="the_username"
                  />
                </div>
              </div>
              <div
                aria-live="polite"
                class="rmd-form-message rmd-form-message--error rmd-form-message--outline"
                id="username--message"
              >
                <p
                  class="rmd-form-message__message"
                  id="username--message-message"
                >
                  The username is already in use: the_username
                </p>
              </div>
              <div
                class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--outline-left rmd-text-field-container--outline-right rmd-password rmd-password--offset"
              >
                <span
                  class="rmd-text-field-addon rmd-text-field-addon--presentational"
                >
                  <span
                    class="material-symbols"
                  >
                    password
                  </span>
                </span>
                <label
                  class="rmd-label rmd-floating-label rmd-floating-label--active rmd-floating-label--inactive"
                  for="password"
                >
                  Password
                </label>
                <input
                  autocomplete="new-password"
                  class="rmd-text-field rmd-text-field--floating rmd-password__input rmd-password__input--offset"
                  id="password"
                  maxlength="64"
                  minlength="6"
                  name="password"
                  required=""
                  type="password"
                  value="password"
                />
                <button
                  aria-label="Show password"
                  aria-pressed="false"
                  class="rmd-button rmd-button--icon rmd-password__toggle"
                  id="password-password-toggle"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="rmd-icon rmd-icon--font material-icons"
                  >
                    remove_red_eye
                  </i>
                  <span
                    class="rmd-ripple-container"
                  />
                </button>
              </div>
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label"
                  style="flex-grow: 1;"
                >
                  <label
                    class="rmd-label rmd-floating-label rmd-floating-label--active rmd-floating-label--inactive"
                    for="long-name"
                  >
                    Full Name
                  </label>
                  <input
                    autocomplete="name"
                    class="rmd-text-field rmd-text-field--floating"
                    id="long-name"
                    maxlength="64"
                    name="longName"
                    required=""
                    type="text"
                    value="Testius Maximus IV"
                  />
                </div>
              </div>
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label"
                  style="flex-grow: 1;"
                >
                  <label
                    class="rmd-label rmd-floating-label rmd-floating-label--active rmd-floating-label--inactive"
                    for="short-name"
                  >
                    First Name
                  </label>
                  <input
                    autocomplete="given-name"
                    class="rmd-text-field rmd-text-field--floating"
                    id="short-name"
                    maxlength="32"
                    name="shortName"
                    type="text"
                    value="Testy"
                  />
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="does-accept-terms"
                      name="doesAcceptTerms"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="does-accept-terms"
                  >
                    <div>
                      I have read and agree to the
                       
                      <a
                        class="text-link"
                        href="https://docs.howdju.com/policies/user-agreement"
                        rel="noreferrer"
                        target="_blank"
                      >
                        User Agreement
                      </a>
                       
                      and the
                       
                      <a
                        class="text-link"
                        href="https://docs.howdju.com/policies/privacy-policy"
                        rel="noreferrer"
                        target="_blank"
                      >
                        Privacy Policy
                      </a>
                      .
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="is-13-years-or-older"
                      name="is13YearsOrOlder"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="is-13-years-or-older"
                  >
                    <div>
                      I am 13 years old or older.
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="has-majority-consent"
                      name="hasMajorityConsent"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="has-majority-consent"
                  >
                    <div>
                      I am old enough in my local jurisdiction to enter into legal agreements and to consent to the processing of my personal data.
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="is-not-gdpr"
                      name="isNotGdpr"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="is-not-gdpr"
                  >
                    <div>
                      I am not located in the European Union (EU), the European Economic Area (EEA), or in any other jurisdiction that is subject to the General Data Protection Regulation (GDPR). (See
                       
                      <a
                        class="text-link"
                        href="https://github.com/Howdju/howdju/issues/24"
                        rel="noreferrer"
                        target="_blank"
                      >
                        Github issue #24
                      </a>
                      .)
                    </div>
                  </label>
                </div>
              </div>
            </div>
            <div
              class="rmd-card__actions"
            >
              <button
                class="rmd-button rmd-button--text rmd-button--text-primary"
                type="button"
              >
                Cancel
                <span
                  class="rmd-ripple-container"
                />
              </button>
              <button
                class="rmd-button rmd-button--text rmd-button--contained rmd-button--primary"
                title="Submit"
                type="submit"
              >
                Complete Registration
                <span
                  class="rmd-ripple-container"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RegistrationConfirmationPage shows error message for missing registration code 1`] = `
<div>
  <div
    id="register-page"
  >
    <div
      class="rmd-grid"
      style="--rmd-grid-cols: 1;"
    >
      <div
        class="rmd-card rmd-card--shadowed rmd-grid__cell"
      >
        <header
          class="rmd-card__header rmd-card__header--center"
        >
          <span
            class="rmd-card__header-content"
          >
            <h5
              class="rmd-card__title"
            >
              Complete Registration
            </h5>
          </span>
        </header>
        <div
          class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
        >
          <span
            class="error-message"
          >
            Missing registration code
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RegistrationConfirmationPage shows form 1`] = `
<div>
  <div
    id="register-page"
  >
    <div
      class="rmd-grid"
      style="--rmd-grid-cols: 1;"
    >
      <div
        class="rmd-card rmd-card--shadowed rmd-grid__cell"
      >
        <header
          class="rmd-card__header rmd-card__header--center"
        >
          <span
            class="rmd-card__header-content"
          >
            <h5
              class="rmd-card__title"
            >
              Complete Registration
            </h5>
          </span>
        </header>
        <div
          class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
        >
          <div
            class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
          >
            Please enter the following to complete your registration
          </div>
          <form>
            <div
              class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
            >
              <ul
                class="error-message"
              />
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--disabled rmd-text-field-container--label rmd-text-field-container--outline-left"
                  style="flex-grow: 1;"
                >
                  <span
                    class="rmd-text-field-addon rmd-text-field-addon--presentational"
                  >
                    <span
                      class="material-symbols"
                    >
                      alternate_email
                    </span>
                  </span>
                  <label
                    class="rmd-label rmd-label--disabled rmd-floating-label rmd-floating-label--active"
                    for="email"
                  >
                    Email
                  </label>
                  <input
                    autocomplete="email"
                    class="rmd-text-field rmd-text-field--floating"
                    disabled=""
                    id="email"
                    type="email"
                    value="the-email@domain.com"
                  />
                </div>
              </div>
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--outline-left"
                  style="flex-grow: 1;"
                >
                  <span
                    class="rmd-text-field-addon rmd-text-field-addon--presentational"
                  >
                    <span
                      class="material-symbols"
                    >
                      person
                    </span>
                  </span>
                  <label
                    class="rmd-label rmd-floating-label"
                    for="username"
                  >
                    Username
                  </label>
                  <input
                    autocomplete="username"
                    class="rmd-text-field rmd-text-field--floating"
                    id="username"
                    maxlength="64"
                    minlength="3"
                    name="username"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label rmd-text-field-container--outline-left rmd-text-field-container--outline-right rmd-password rmd-password--offset"
              >
                <span
                  class="rmd-text-field-addon rmd-text-field-addon--presentational"
                >
                  <span
                    class="material-symbols"
                  >
                    password
                  </span>
                </span>
                <label
                  class="rmd-label rmd-floating-label"
                  for="password"
                >
                  Password
                </label>
                <input
                  autocomplete="new-password"
                  class="rmd-text-field rmd-text-field--floating rmd-password__input rmd-password__input--offset"
                  id="password"
                  maxlength="64"
                  minlength="6"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
                <button
                  aria-label="Show password"
                  aria-pressed="false"
                  class="rmd-button rmd-button--icon rmd-password__toggle"
                  id="password-password-toggle"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="rmd-icon rmd-icon--font material-icons"
                  >
                    remove_red_eye
                  </i>
                  <span
                    class="rmd-ripple-container"
                  />
                </button>
              </div>
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label"
                  style="flex-grow: 1;"
                >
                  <label
                    class="rmd-label rmd-floating-label"
                    for="long-name"
                  >
                    Full Name
                  </label>
                  <input
                    autocomplete="name"
                    class="rmd-text-field rmd-text-field--floating"
                    id="long-name"
                    maxlength="64"
                    name="longName"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                style="display: flex;"
              >
                <div
                  class="rmd-text-field-container rmd-text-field-container--outline rmd-text-field-container--hoverable rmd-text-field-container--label"
                  style="flex-grow: 1;"
                >
                  <label
                    class="rmd-label rmd-floating-label"
                    for="short-name"
                  >
                    First Name
                  </label>
                  <input
                    autocomplete="given-name"
                    class="rmd-text-field rmd-text-field--floating"
                    id="short-name"
                    maxlength="32"
                    name="shortName"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="does-accept-terms"
                      name="doesAcceptTerms"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="does-accept-terms"
                  >
                    <div>
                      I have read and agree to the
                       
                      <a
                        class="text-link"
                        href="https://docs.howdju.com/policies/user-agreement"
                        rel="noreferrer"
                        target="_blank"
                      >
                        User Agreement
                      </a>
                       
                      and the
                       
                      <a
                        class="text-link"
                        href="https://docs.howdju.com/policies/privacy-policy"
                        rel="noreferrer"
                        target="_blank"
                      >
                        Privacy Policy
                      </a>
                      .
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="is-13-years-or-older"
                      name="is13YearsOrOlder"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="is-13-years-or-older"
                  >
                    <div>
                      I am 13 years old or older.
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="has-majority-consent"
                      name="hasMajorityConsent"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="has-majority-consent"
                  >
                    <div>
                      I am old enough in my local jurisdiction to enter into legal agreements and to consent to the processing of my personal data.
                    </div>
                  </label>
                </div>
              </div>
              <div
                class="howdju-checkbox"
              >
                <div
                  class="rmd-toggle-container"
                >
                  <span
                    class="rmd-toggle"
                  >
                    <input
                      class="rmd-toggle__input"
                      id="is-not-gdpr"
                      name="isNotGdpr"
                      type="checkbox"
                    />
                    <span
                      class="rmd-toggle__icon rmd-toggle__icon--overlay"
                    >
                      <i
                        aria-hidden="true"
                        class="rmd-icon rmd-icon--font material-icons"
                      >
                        check_box
                      </i>
                    </span>
                    <span
                      class="rmd-ripple-container"
                    />
                  </span>
                  <label
                    class="rmd-label"
                    for="is-not-gdpr"
                  >
                    <div>
                      I am not located in the European Union (EU), the European Economic Area (EEA), or in any other jurisdiction that is subject to the General Data Protection Regulation (GDPR). (See
                       
                      <a
                        class="text-link"
                        href="https://github.com/Howdju/howdju/issues/24"
                        rel="noreferrer"
                        target="_blank"
                      >
                        Github issue #24
                      </a>
                      .)
                    </div>
                  </label>
                </div>
              </div>
            </div>
            <div
              class="rmd-card__actions"
            >
              <button
                class="rmd-button rmd-button--text rmd-button--text-primary"
                type="button"
              >
                Cancel
                <span
                  class="rmd-ripple-container"
                />
              </button>
              <button
                aria-disabled="true"
                class="rmd-button rmd-button--text rmd-button--contained rmd-button--disabled"
                title="Please complete the form to continue"
                type="submit"
              >
                Complete Registration
                <span
                  class="rmd-ripple-container"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RegistrationConfirmationPage shows success message upon confirmation. 1`] = `
<div>
  <div
    id="register-page"
  >
    <div
      class="rmd-grid"
      style="--rmd-grid-cols: 1;"
    >
      <div
        class="rmd-card rmd-card--shadowed rmd-grid__cell"
      >
        <header
          class="rmd-card__header rmd-card__header--center"
        >
          <span
            class="rmd-card__header-content"
          >
            <h5
              class="rmd-card__title"
            >
              Complete Registration
            </h5>
          </span>
        </header>
        <div
          class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
        >
          <div
            class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin"
          >
            You are now logged in.
          </div>
          <div
            class="rmd-card__actions"
          >
            <button
              class="rmd-button rmd-button--text rmd-button--contained rmd-button--primary"
              href="/recent-activity"
              type="button"
            >
              Go to recent activity
              <span
                class="rmd-ripple-container"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
